client/engine/engine.go

```diff
@@ -0,0 +1,135 @@
+// Package engine contains the types and imperative code for the business logic of a go-nitro Client.
+package engine // import "github.com/statechannels/go-nitro/client/engine"
+
+import (
+	"math/big"
+
+	"github.com/statechannels/go-nitro/channel/state"
+	"github.com/statechannels/go-nitro/client/engine/store"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+// Engine is the imperative part of the core business logic of a go-nitro Client
+type Engine struct {
+	// inbound go channels
+	API   chan APIEvent
+	Chain chan ChainEvent
+	Inbox chan Message
+
+	store store.Store // A Store foe persisting important data
+}
+
+// APIEvent is an internal representation of an API call
+type APIEvent struct {
+	ObjectiveToSpawn   protocols.Objective   // try this first
+	ObjectiveToReject  protocols.ObjectiveId // then this
+	ObjectiveToApprove protocols.ObjectiveId // then this
```

Comment:
Here I'm trying out one way to work around the lack of union types in Go. Instead of a union of SpawnObjective, RejectObjective and ApproveObjective, here I'm using an intersection of those types. With this pattern, we expect the types in the intersection to be discriminated by the APIEvent having exactly one non-null field. Futher, we check for nullity in the order shown. I'm undecided about this pattern in general.

Some alternatives:

1. We could allow multiple non-null fields and handle them all. So the logic inside the handler would be `if..then` rather than `switch` (i.e. it wouldn't be mutually exclusive). This would be make the behaviour less surprising, given we are using an intersection type.

2. Another alternative is to replace this type with an empty interface, and then have a type switch inside `handleAPIEvent(event interface{})`. This isn't type safe, though (nothing stops you piping garbage into that handler).

3. We know of some other ways to emulate type unions.